docs/proposals/predictive-horizontal-pod-autoscaling.md

```diff
@@ -0,0 +1,232 @@
+```
+
+Predictive horizontal autoscaling will use a similar algorithm with one
+significant difference. Rather than calculating based on the
+`CurrentPodsResourceConsumption`, we will calculate based on
+`FuturePodsResourceConsumption`, which we initially define as
+`FuturePodsResourceConsumption = CurrentPodResourceConsumption +
+(PodsAverageInitializationTime * DerivativeCurrentPodsResourceConsumptionPerSecond)`.
+The calculation for `FuturePodsResourceConsumption` could become more complex,
+incorporating increasingly complex methods of prediction, in the future.
+We now calculate `TargetNumOfPods` with the following:
+
+```
+TargetNumOfPods = ceil(sum(FuturePodsResourceConsumption)) / Target
+```
+
+The modifications for this option will predominantly occur in
+`pkg/controllers/podautoscaler/horizontal.go` in the
+`computeReplicasForCPUUtilization` function. The new algorithm will be used
+only if predictive auto-scaling is turned on (as discussed in the next
+section).
+
+## Implementation
+
+### Turning on Prediction
+
+Initially, the configuration value for enabling predictive autoscaling will be recorded in
+the `HorizontalPodAutoscaler.ObjectMeta.Annotations`.
+
+This annotation can be done via the following command:
+
+```sh
+kubectl annotate hpa foo predictive=’true’
+```
+
+### Other modifications
+
+Additionally, the algorithm for predictive autoscaling requires two new pieces
+of information: a pod’s initialization time and the rate of change for the
+metric upon which scaling behavior is determined (currently only CPU). These
+two values are currently untracked; the following two sections describe the
+changes to record them.
+
+#### Recording `AveragePodInitializationTime`
+
+`AveragePodInitializeTime` has a couple of important nuances. First, it is not
+simply the amount of time between when a pod is ordered to be created and when
+it is created. Rather, we want to record the amount of time from when a pod is
+ordered to be created and when it is ready to perform the task for which it is
+created. For example, a pod may be created, but if the purpose of the pod is to
+run a webserver, and that web server takes five minutes to be able to serve
+requests, then the mere fact that the pod is created does not help balance the
+load nor does it accomplish the purpose of autoscaling. It is the difference
+between the pod’s creation time and the time at which the pod is ready to
+perform its task that we are interested in; we call this metric
+`InitializationTime`.
+
+We will calculate `AveragePodInitializationTime` by storing a measure of
+`InitializationTime` for each pod created during auto-scaling in the pod’s
+`Annotations`. We can then use the pod selector specified by the horizontal pod
+auto-scaler to retrieve and average these values, resulting in the
+`AverageInitializationTime` metric.
+
+We ensure each newly created pod tracks `InitializationTime` in
```

Comment:
Potentially you don't need to have a watch. You can take the last pod condition time change and calculate the difference between pod creation startup: 
```
"status": {
                "phase": "Running",
                "conditions": [
                    {
                        "type": "Ready",
                        "status": "True",
                        "lastProbeTime": null,
                        "lastTransitionTime": "2016-03-04T12:25:49Z"
                    }
                ],
```
Pod creation:
```
 "creationTimestamp": "2016-03-04T12:25:01Z",
```
In my case initTime was 49 seconds.